lib/rust/parser/debug/tests/parse.rs

```diff
@@ -698,21 +700,34 @@ fn unevaluated_argument() {
 
 #[test]
 fn unary_operator_missing_operand() {
-    test_invalid("main ~ = x");
+    expect_invalid_node("main ~ = x");
 }
 
 #[test]
 fn unary_operator_at_end_of_expression() {
-    test_invalid("foo ~");
+    expect_invalid_node("foo ~");
 }
 
 #[test]
-fn plus_negative() {
-    let code = ["x = x+-x"];
-    let expected = block![
-        (Assignment (Ident x) "=" (OprApp (Ident x) (Ok "+") (UnaryOprApp "-" (Ident x))))
+fn unspaced_operator_sequence() {
+    let cases = [
+        // Add a negated value.
+        ("x = y+-z", block![
+            (Assignment (Ident x) "=" (OprApp (Ident y) (Ok "+") (UnaryOprApp "-" (Ident z))))]),
+        // Create an operator section that adds a negated value to its input.
+        ("x = +-z", block![
+            (Assignment (Ident x) "=" (OprSectionBoundary 1
+                (OprApp () (Ok "+") (UnaryOprApp "-" (Ident z)))))]),
+        // Create an operator section that adds its input, negated, to a value.
+        ("x = y+-", block![
+            (Assignment (Ident x) "=" (OprSectionBoundary 1
+                (OprApp (Ident y) (Ok "+") (UnaryOprApp "-" ()))))]),
+        // Assign a negative number to x.
+        ("x=-1", block![(Assignment (Ident x) "=" (UnaryOprApp "-" (Number () "1" ())))]),
+        // Assign a negated value to x.
+        ("x=-y", block![(Assignment (Ident x) "=" (UnaryOprApp "-" (Ident y)))]),
     ];
-    test(&code.join("\n"), expected);
+    cases.into_iter().for_each(|(code, expected)| test(code, expected));
```

Comment:
It is easier to read. The tests are organized into Section -> Subsection -> Function -> Cases; the Function is an important unit because it tests one aspect of the functionality, and is the lowest level that can be usefully named. It's true that this is less convenient for debugging when a test fails. Sometimes I write a function or macro that runs all the cases unconditionally and reports which failed, while allowing them to be defined together without intervening per-testcase boilerplate (I have used this approach in the new fuzzy matcher). I have not bothered to do so for the parser test cases because they have rarely failed, so my main priority for the test case API has been for the tests to serve as readable documentation of the edge cases.